src/metabase/models/card.clj

```diff
@@ -274,6 +274,46 @@
   pre-update-check-sandbox-constraints
   (atom identity))
 
+(defn- update-parameters-using-card-as-values-sources
+  "Update the config of parameter on any Dashboard/Card use this `card` as values source .
+
+  Remove parameter.values_source_type and set parameter.values_source_type to nil ( the default type ) when:
+  - card is archived
+  - card.result_metadata changes"
+  [{id :id, :as changes}]
+  (let [parameter-cards   (db/select ParameterCard :card_id id)]
+    (doseq [[[po-type po-id] param-cards]
+            (group-by (juxt :parameterized_object_type :parameterized_object_id) parameter-cards)]
+      (let [model                  (case po-type :card 'Card :dashboard 'Dashboard)
+            {:keys [parameters]}   (db/select-one [model :parameters] :id po-id)
+            affected-param-ids-set (cond
+                                     ;; update all parameter that use this card as source
+                                     (:archived changes)
+                                     (set (map :parameter_id param-cards))
+
+                                     ;; update only parameter that has value_field no longer in this card
+                                     (:result_metadata changes)
+                                     (let [param-id->parameter (m/index-by :id parameters)]
+                                       (->> (for [param-card param-cards]
+                                              ;; if cant find the value-field in result_metadata, then we should remove it
+                                              (when (nil? (qp.util/field->field-info
+                                                            (get-in (param-id->parameter (:parameter_id param-card)) [:values_source_config :value_field])
+                                                            (:result_metadata changes)))
+                                                (:parameter_id param-card)))
+                                            (filter some?)
+                                            set))
+
+                                     :else #{})
+            new-parameters (map (fn [parameter]
+                                  (if (affected-param-ids-set (:id parameter))
+                                    (-> parameter
+                                        (assoc :values_source_type nil)
+                                        (dissoc :values_source_config))
```

Comment:
Is this equivalent to `(dissoc :values_source_config :values_source_type)`? Or does `nil` values_source_type have a special meaning / is required by the schema

Reply:
Okay. It's pretty rare to see `(assoc x nil)` so I thought it might be significant to the runtime behaviour. But I get the intent